File: JS/BOJ-S5-2828-Apple_gathering_game.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n')
  .map((v) => v.trimEnd());

let output = [];
const solution = (input) => {
  const [N, M] = input[0].split(' ').map(Number);
  const arr = input.slice(2).map(Number);

  let l = 1;
  let ret = 0;

  arr.forEach((v) => {
    let r = l + M - 1;

    if (l <= v && v <= r) return;

    if (v < l) {
      ret += l - v;
      l = v;
    } else if (v >= l) {
      l += v - r;
      ret += v - r;
    }
  });

  output.push(ret);
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join('').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join('').trimEnd(),
//   (err) => console.error
// );

File: CPP/BOJ-S5-2828-Apple_gathering_game.cpp
#include <bits/stdc++.h>
using namespace std;

int m, n, j, l, r, tmp, ret;

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  // freopen("input.txt", "r", stdin);
  // freopen("output.txt", "w", stdout);

  cin >> n >> m >> j;

  l = 1;
  for (int i = 0; i < j; ++i)
  {
    r = l + m - 1;
    cin >> tmp;
    if (tmp >= l && tmp <= r)
      continue;

    if (tmp < l)
    {
      ret += (l - tmp);
      l = tmp;
    }
    else if (tmp >= l)
    {
      l += (tmp - r);
      ret += (tmp - r);
    }
  }

  cout << ret << "\n";

  return 0;
}